modules/transformation.py

```diff
@@ -0,0 +1,609 @@
+                        exc_info=True,
+                    )
+
+            ret_request.update({"expAge": int(max(exp_age, 0))})
+
+            # NOW add all the headers from both the request and response.
+            ret_request.update({k: ", ".join(v) for k, v in request_headers.items()})
+
+            # TODO consider doing this sooner? why process if status is bad?
+            # wrap it up
+            first_req = False
+            first_html = False
+            if not first_url:
+                if (400 <= status <= 599) or 12000 <= status:
+                    logging.error(
+                        "ERROR($gPagesTable pageid: {}): The first request ({}}) failed with status {}}.",
+                        pageid,
+                        url,
+                        status,
+                    )
+                    return None
+                # This is the first URL found associated with the page - assume it's the base URL.
+                first_req = True
+                first_url = url
+
+            if not first_html_url:
+                # This is the first URL found associated with the page that's HTML.
+                first_html = True
+                first_html_url = url
+
+            ret_request.update({"firstReq": first_req, "firstHtml": first_html})
+
+            requests.append(ret_request)
+
+        return requests, first_url, first_html_url
+
+    @staticmethod
+    def import_page(page, status_info):
+        return {
+            "client": status_info["client"],
+            "date": status_info["date"],
+            "pageid": status_info["pageid"],
+            "createDate": int(datetime.datetime.now().timestamp()),
+            "startedDateTime": utils.datetime_to_epoch(
+                page["startedDateTime"], status_info
+            ),
+            "archive": status_info["archive"],
+            "label": status_info["label"],
+            "crawlid": status_info["crawlid"],
+            # TODO confirm - it's ok to get url from page info rather than status info as originally implemented?
+            "url": page["_URL"],
+            "urlhash": utils.get_url_hash(page["_URL"]),
+            "urlShort": page["_URL"][:255],
+            "TTFB": page.get("_TTFB"),
+            "renderStart": page.get("_render"),
+            "fullyLoaded": page.get("_fullyLoaded"),
+            "visualComplete": page.get("_visualComplete"),
+            "onLoad": page.get("_docTime")
+            if page.get("_docTime") != 0
+            else max(page.get("_visualComplete"), page.get("_fullyLoaded")),
+            "gzipTotal": page.get("_gzip_total"),
+            "gzipSavings": page.get("_gzip_savings"),
+            "numDomElements": page.get("_domElements"),
+            "onContentLoaded": page.get("_domContentLoadedEventStart"),
+            "cdn": page.get("_base_page_cdn"),
+            "SpeedIndex": page.get("_SpeedIndex"),
+            "PageSpeed": page.get("_pageSpeed", {}).get("score"),
```

Comment:
@pmeenan AFAICT this field never existed in the HAR

Reply:
Thanks for confirming. I think this code is fine as-is and we can look into adding better summary data to the tables in https://github.com/HTTPArchive/data-pipeline/issues/15.